src/ir/auth_logic/universe_relation_insertion.h

```diff
@@ -0,0 +1,449 @@
+    elements to universes.
+    2) extending the program with facts that use delegation to pass membership
+    of universes across principals.
+
+This translation is a compiler pass that operates on the data structure
+described in `ast.h` and produces a new instance of the data structure in
+`ast.h`. It should come before the translation in `lowering_ast_datalog.h`.
+*/
+
+#ifndef SRC_IR_AUTH_LOGIC_UNIVERSE_RELATION_INSERTION_H_
+#define SRC_IR_AUTH_LOGIC_UNIVERSE_RELATION_INSERTION_H_
+
+#include "absl/container/flat_hash_map.h"
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+#include "src/ir/auth_logic/ast.h"
+#include "src/ir/auth_logic/auth_logic_ast_traversing_visitor.h"
+#include "src/ir/datalog/program.h"
+
+namespace raksha::ir::auth_logic {
+
+class UniverseRelationInsertion {
+ public:
+  // A map from the names of principal / predicate argument literals
+  // (the constants wrapped in quotes) to typings.
+  using TypeEnv = absl::flat_hash_map<std::string, datalog::ArgumentType>;
+  // A map from predicate names to predicate declarations.
+  using DeclEnv =
+      absl::flat_hash_map<std::string, datalog::RelationDeclaration>;
+
+  static Program InsertUniverseRelations(const Program& prog) {
+    // Generate relation declaration environment
+    RelationDeclarationEnvironmentVisitor rel_visitor;
+    prog.Accept(rel_visitor);
+    DeclEnv decl_env = rel_visitor.decl_env();
+    // Generate constant type environment
+    TypeEnvironmentGenerationVisitor type_env_visitor(decl_env);
+    prog.Accept(type_env_visitor);
+    TypeEnv type_env = type_env_visitor.type_env();
+    // Generate universe populating facts
+    std::vector<SaysAssertion> new_says_assertions =
+        GetUniverseDefiningFacts(type_env);
+    // Transform assertions to include conditions
+    // that check the universes
+    UniverseConditionTransformer universe_condition_transformer(decl_env);
+    for (const auto& says_assertion : prog.says_assertions()) {
+      new_says_assertions.push_back(
+          universe_condition_transformer.AddUniverseConditions(says_assertion));
+    }
+
+    // Add declarations for universe relations.
+    auto universe_declarations = GetUniverseDeclarations(prog);
+    auto new_declarations = prog.relation_declarations();
+    new_declarations.insert(new_declarations.end(),
+                            universe_declarations.begin(),
+                            universe_declarations.end());
+
+    return Program(new_declarations, new_says_assertions, prog.queries());
+  }
+
+ private:
+  static std::string TypeToUniverseName(datalog::ArgumentType arg_type) {
+    switch (arg_type.kind()) {
+      case datalog::ArgumentType::Kind::kNumber:
+        return "isNumber";
+      case datalog::ArgumentType::Kind::kPrincipal:
+        return "isPrincipal";
+      case datalog::ArgumentType::Kind::kCustom:
+        return absl::StrCat("is", arg_type.name());
+    }
```

Comment:
With switch we have one less vtable lookup, right?

Reply:
Yeah, but that's the kind of thing you should only care about if you're getting complaints about code being too slow and your profiler points to this happening over and over again.

A single vtable lookup doesn't matter for time complexity and at a high level, time complexity is the only thing that matters (well, space complexity too, but that's usually harder to make too much of a mess with). Looking a tiny bit more closely, you can help performance out without much additional effort or readability impact with some well known patterns, like "use offline sorts instead of online sorts when possible" or "use densely allocated structures rather than pointer-heavy structures when possible" or "reserve your space in your vector in advance when possible instead of leaning on `push_back`". But even those matter only sometimes.

Reply:
On second thought, I think this is best left in the universe relation pass since at the time of writing the concept of a universe lives exclusively in this pass, so I think it is best to also keep this function about universes there? we can also always revise if universes become more of a first-class citizen for some reason. Feel free to re-open if you object, of course.

Reply:
OK, that's reasonable.